RFC.md

```diff
@@ -1,12 +1,289 @@
+
+- If all the indices are equal (that is, `idxA_1 == idxB_1 && ... && idxA_N == idxB_N`), 
+  then the resulting pointer values are equal.
+
+- If any index value is not equal (that is, there exists an `i` such that `idxA_i != idxB_i`),
+  then the resulting pointer values are not equal.
+
+
+##### Address computation:
+Consider an invocation of `multidim_array_index`:
+
+```
+<result> = multidim_array_index <ty> <ty>* <ptrval> <str_0>, <idx_0>, <str_1> <idx_1>, ..., <str_n> <idx_n>
+```
+
+If the pairs are denoted by `(str_i, idx_i)`, where `str_i` denotes the stride
+and `idx_i` denotes the index of the ith pair, then the final address (in bytes)
+is computed as:
+
+```
+ptrval + len(ty) * [(str_0 * idx_0) + (str_1 * idx_1) + ... (str_n * idx_n)]
+```
+
+## Transitioning to `multidim_array_index`: Allow `multidim_array_index` to refer to a GEP instruction:
+
+
+##### Choice 1: Write a `multidim_array_index` to `GEP` pass, with the `GEP` annotated with metadata
+
+This pass will flatten all `multidim_array_index` expressions into a `GEP` annotated with metadata. This metadata will indicate that the index expression computed by the lowered GEP is guaranteed to be in a canonical form which allows the analysis
+to infer stride and index sizes.
+
+A multidim index of the form:
+```
+%arrayidx = multidim_array_index i64 i64* %A, %str_1, %idx_1, %str_2, %idx_2
+```
+
+is lowered to:
+
+```
+%mul1 = mul nsw i64 %str_1, %idx_1
+%mul2 = mul1 nsw i64 %str_2, %idx_2
+%total = add nsw i64 %mul2, %mul1
+%arrayidx = getelementptr inbounds i64, i64* %A, i64 %total, !multidim !1
+```
+with guarantees that the first term in each multiplication is the stride
+and the second term in each multiplication is the index. (What happens
+if intermediate transformation passes decide to change the order? This seems
+complicated).
+
+**TODO:** Lookup how to attach metadata such that the metadata can communicate
+which of the values are strides and which are indeces
+
+
+
+# Caveats
+
+Currently, we assume that the array shape is immutable. However, we will need to deal with 
+being able to express `reshape` like primitives where the array shape can be mutated. However,
+this appears to make expressing this information quite difficult: We now need to attach the shape
+information to an array per "shape-live-range". 
+
+
+## Appendix-A
+
+##### A realistic, more involved example of dependence analysis going wrong
+
+```cpp
+// In an array A of size (n0 x n1), 
+// fill a subarray of size (s0 x s1)
+// which starts at an offset (o0 x o1) in the larger array A
+void set_subarray(unsigned n0, unsigned n1,
+	unsigned o0, unsigned o1,
+	unsigned s0, unsigned s1,
+	float A[n0][n1]) {
+	for (unsigned i = 0; i < s0; i++)
+		for (unsigned j = 0; j < s1; j++)
+				S: A[i + o0][j + o1] = 1;
+}
+```
+We first reduce this index expression to a sum of products:
+
+```
+(i + o0) * n0 + (j + o1) = n0i + n0o0 + j + o1
```

Comment:
Why is this `* n0`?  Shouldn't it be `* n1`?

Reply:
Indeed, thanks!